File: core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.entity;

import java.io.File;
import java.io.Serializable;
import java.util.LinkedList;
import org.simpleframework.xml.Attribute;
import org.simpleframework.xml.ElementList;
import org.simpleframework.xml.Root;

@Root(name = "library", strict = false)
public class LibraryNetworkEntity {

  @ElementList(name = "book", inline = true, required = false)
  public LinkedList<Book> book;

  @Attribute(name = "version", required = false)
  private String version;

  public LinkedList<Book> getBooks() {
    return this.book;
  }

  public String getVersion() {
    return this.version;
  }

  @Root(name = "book", strict = false)
  public static class Book implements Serializable {

    @Attribute(name = "id", required = false)
    public String id;

    @Attribute(name = "title", required = false)
    public String title;

    @Attribute(name = "description", required = false)
    public String description;

    @Attribute(name = "language", required = false)
    public String language;

    @Attribute(name = "creator", required = false)
    public String creator;

    @Attribute(name = "publisher", required = false)
    public String publisher;

    @Attribute(name = "favicon", required = false)
    public String favicon;

    @Attribute(name = "faviconMimeType", required = false)
    public String faviconMimeType;

    @Attribute(name = "date", required = false)
    public String date;

    @Attribute(name = "url", required = false)
    public String url;

    @Attribute(name = "articleCount", required = false)
    public String articleCount;

    @Attribute(name = "mediaCount", required = false)
    public String mediaCount;

    @Attribute(name = "size", required = false)
    public String size;

    @Attribute(name = "name", required = false)
    public String bookName;

    @Attribute(name = "tags", required = false)
    public String tags;

    public int searchMatches = 0;
    @Deprecated
    public File file;
    @Deprecated
    public String remoteUrl;

    public String getId() {
      return this.id;
    }

    public String getTitle() {
      return this.title;
    }

    public String getDescription() {
      return this.description;
    }

    public String getLanguage() {
      return this.language;
    }

    public String getCreator() {
      return this.creator;
    }

    public String getPublisher() {
      return this.publisher;
    }

    public String getFavicon() {
      return this.favicon;
    }

    public String getFaviconMimeType() {
      return this.faviconMimeType;
    }

    public String getDate() {
      return this.date;
    }

    public String getUrl() {
      return this.url;
    }

    public String getArticleCount() {
      return this.articleCount;
    }

    public String getMediaCount() {
      return this.mediaCount;
    }

    public String getName() {
      return this.bookName;
    }

    public String getSize() {
      return this.size;
    }

    // Two books are equal if their ids match
    @Override
    public boolean equals(Object obj) {
      if (obj instanceof Book) {
        if (((Book) obj).getId() != null && ((Book) obj).getId().equals(getId())) {
          return true;
        }
      }
      return false;
    }

    // Only use the book's id to generate a hash code
    @Override
    public int hashCode() {
      return getId().hashCode();
    }
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/entity/LibraryNetworkEntity.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.entity

import android.os.Parcelable
import androidx.room.Entity
import androidx.room.TypeConverter
import org.simpleframework.xml.Attribute
import org.simpleframework.xml.ElementList
import org.simpleframework.xml.Root
import java.io.File
import java.io.Serializable
import java.util.LinkedList

@Root(name = "library")
class LibraryNetworkEntity {
  @field:ElementList(name = "book", inline = true, required = false)
  var book: LinkedList<Book>? = null

  @field:Attribute(name = "version", required = false)
  var version: String? = null

  @Root(name = "book", strict = false)
  class Book : Serializable {
    @field:Attribute(name = "id", required = false)
    var id: String = ""

    @field:Attribute(name = "title", required = false)
    var title: String = ""

    @field:Attribute(name = "description", required = false)
    var description: String? = null

    @field:Attribute(name = "language", required = false)
    var language: String = ""

    @field:Attribute(name = "creator", required = false)
    var creator: String = ""

    @field:Attribute(name = "publisher", required = false)
    var publisher: String = ""

    @field:Attribute(name = "favicon", required = false)
    var favicon: String = ""

    @field:Attribute(name = "faviconMimeType", required = false)
    var faviconMimeType: String? = null

    @field:Attribute(name = "date", required = false)
    var date: String = ""

    @field:Attribute(name = "url", required = false)
    var url: String? = null

    @field:Attribute(name = "articleCount", required = false)
    var articleCount: String? = null

    @field:Attribute(name = "mediaCount", required = false)
    var mediaCount: String? = null

    @field:Attribute(name = "size", required = false)
    var size: String = ""

    @field:Attribute(name = "name", required = false)
    var bookName: String? = null

    @field:Attribute(name = "tags", required = false)
    var tags: String? = null
    var searchMatches = 0

    @Deprecated("")
    var file: File? = null

    @Deprecated("")
    var remoteUrl: String? = null

    // Two books are equal if their ids match
    override fun equals(other: Any?): Boolean {
      if (other is Book) {
        if (other.id == id) {
          return true
        }
      }
      return false
    }

    // Only use the book's id to generate a hash code
    override fun hashCode(): Int = id.hashCode()
  }
}
